README.md
﷽

# Residue Node.js Client

[![Build Status](https://img.shields.io/travis/muflihun/residue-node/master.svg)](https://travis-ci.org/muflihun/residue-node/branches)
[![Build status](https://ci.appveyor.com/api/projects/status/rp6ukh5apm5ryjxq?svg=true)](https://ci.appveyor.com/project/abumusamq/residue-node)
[![Version](https://img.shields.io/npm/v/residue.svg)](https://www.npmjs.com/package/residue)
[![GitHub license](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/muflihun/residue-node/blob/master/LICENSE)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/MuflihunDotCom/25)

Secure Node.js library to interact with residue server seamlessly.

This module provides interface for connecting and interacting with residue server seamlessly, means, once you are connected this module takes care of expired tokens and clients and keep itself updated with latest tokens and ping server when needed to stay alive.

## API
#### `connect(options)`
Connects application to residue using params. If options is not specified, you should use `loadConfiguration` to load the options

Valid options are:

```
{
    url: "<host_where_residue_server_is_listening>:<residue_connection_port>",
    application_id: <app_name [optional]>,
    rsa_key_size: <key_size_for_initial_final_key_transmission [optional]>,
    utc_time: <whether_to_use_UTC_time [optional]>,
    time_offset: <time_offset_in_seconds [optional]>,
    client_id: <client_id_that_server_knows_you_as [optional]>,
    client_private_key: <full_path_of_private_key> [must be provided with client_id],
    client_public_key: <full_path_of_public_key> [must be provided with client_id],
    client_key_secret: <base16 encoded secret (passphrase) for encrypted private key if any>,
    server_public_key: <full_path_of_server_public_key>
}
```

Please refer to [`loadConfiguration`](https://muflihun.github.io/residue/docs/class_residue.html#a8292657c93a775b6cbf22c6d4f1166f4) in our C++ library's documentation for more details.

#### `loadConfiguration(jsonFilename)`
Loads configurations / options via json file. Returns true if successfully loaded, otherwise false.

This does not verify the options. Options are validated in `connect()` function

#### `getLogger(logger_id)`
Returns logger class for logging interface

## Usage
```js
const ResidueClient = require('residue');

const residue = new ResidueClient();

const logger = residue.getLogger('sample-app');

const confFile = 'client.conf.json';
if (residue.loadConfiguration(confFile)) {
    residue.connect();
}

// or
// residue.loadConfiguration({ url: ... })
// or
// residue.loadConfiguration('{ url: ... }')

// ALTERNATIVELY
residue.connect({
    url: ...
});

logger.info('simple log');

logger.info('array %s', [1, 2, 3]);

var person = { 'name': 'Adam', 'age': 960, }
logger.info('obj %s', person);

logger.info('null %s', null);

logger.info('undefined %s', undefined);

```

## Log
```js
logger.info(...);
logger.warn(...);
logger.error(...);
logger.debug(...);
logger.trace(...);
logger.fatal(...);
logger.verbose(verbose_level, ...);
```

## Native Binding
Residue Node.js also comes with native binding that uses [C++ client library](https://github.com/muflihun/residue-cpp) to manage connections and asyncronous requests.

If you have installed [`residue-native`](https://www.npmjs.com/package/residue-native) package alongside `residue`, native binding will be used otherwise it will fallback to JS implementation.

## Sample
You can check out [sample client apps](https://github.com/muflihun/residue-node/blob/master/samples) for practical use of this package.

## License
```
Copyright 2017-present Muflihun Labs
Copyright 2017-present @abumusamq

https://github.com/muflihun/
https://muflihun.github.io/
https://muflihun.com/

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Change Log

## [3.0.0]
- ResidueClient constructible to create unique instances for various clients

## [2.3.18]
- Fixed warning log

## [2.3.17]
- Fixed translating Error

## [2.3.13]
- Fixed fatal log for native
- Register logger explicitely

## [2.3.11]
- Introduced native binding
- Added `version()` and `type()`

## [2.1.0]
- Private key uses hex

## [2.0.1]
- Compatibility with server 2.0.0

## [1.2.5]
- Removed plain request

## [1.2.4]
 - `loadConfiguration` now accepts object

## [1.2.1]
 - Included folders updated

## [1.2.0]
### Updates
- Support formatted logging

## [1.1.5]
### Fixes
- Do not send multiple touch requests when one is pending
- Crash when client is expired (#24)

### Updates
- License information update

## [1.1.4] - 06-10-2017
### Changes
- Display server version when connected

## [1.1.2] - 16-08-2017
### Changes
- Crypto API check
- Stable release

## [1.0.0-beta.16] - 15-08-2017
### Changes
- License added to the top of the file

## [1.0.0-beta.15] - 08-08-2017
### Changes
- Strict mode fix

## [1.0.0-beta.14] - 08-08-2017
### Changes
- Changed verbose signature from `msg, level` to `level, msg`
- Updated to use `requires_token` naming

## [1.0.0-beta.13] - 03-08-2017
### Update
- Changed licence to Apache 2.0

## [1.0.0-beta.11] - 03-08-2017
### Fixed
- Use of `CHECK_TOKENS` server flag to reduce overhead of pulling token when not needed
- Change ping to touch
- Re-estabilish connection if remote disconnected and dispatch all the logs
- Updated touch threshold to 2 minutes

## [1.0.0-beta.10]
### Fixed
- Fixed issue with token retrieving when client is dead
- Fixed issue with pinging client when `client_age` < 60
- Fixed issue with key when client created using different key size
- Fixed compression flag
- Fixed issue with token retrieving using default access code

## [1.0.0-beta.9]
### Fixed
- Use `path` module to resolve relative file paths

### Updates
- Use `crypto` module for RSA operations
- Ability to use encrypted private key
- Ability to specify public key for encrypted private keys

## [1.0.0-beta.7]
### Fixed
- Fixed log for generating key size
- Fixed issues with getting correct filename, line number and function name

### Added
- Added `loadConfiguration` to load configurations from json file

## [1.0.0-beta.6]
### Fixed
- Obtaining token race-condition issue #26
- Dead client not reconnecting #27
- Fixed `shouldSendPing()` calculations
- Fixed issue with verifying token with `0` life

### Changes
- Moved to separate repo
- Updated API to use configurations as C++ API

## [1.0.0-beta] - 31-03-2017
### Added
- Send server flags with final connection #14
- Support sending plain log requests in lib #13
- Support compression #19

### Changes
- Removed using `ripe` as backup
- Replaced `Logger` with `getLogger`

## [1.0.0-alpha] - 19-03-2017
### Changes
 - Initial release


src/residue.js
//
// Official Node.js client library for Residue logging server
//
// Copyright 2017-present Muflihun Labs
//
// This module provides interface for connecting and interacting with
// residue server seamlessly. Once you are connected this module
// takes care of lost connections, expired clients
// and keep itself updated with parameters and touch server when
// needed to stay alive.
//
// Author: @abumusamq
//
// https://muflihun.com
// https://muflihun.github.io/residue
// https://github.com/muflihun/residue-node
//

"use strict";

const fs = require('fs');
const path = require('path');
const net = require('net');
const util = require('util');
const zlib = require('zlib');
const NodeRSA = require('node-rsa');
const CommonUtils = require('residue-utils');
let crypto;
try {
  crypto = require('crypto');
} catch (err) {
  console.log('residue package requires crypto (https://nodejs.org/api/crypto.html). It is disabled in your version of node!');
}

const DEBUGGING = false;
const VERBOSE_LEVEL = 0;

// Various connection types accepted by the server
const ConnectType = {
  CONN: 1,
  ACK: 2,
  TOUCH: 3,
};

const Flag = {
  NONE: 0,
  ALLOW_UNKNOWN_LOGGERS: 1,
  ALLOW_BULK_LOG_REQUEST: 16,
  COMPRESSION: 256
};

const PACKET_DELIMITER = '\r\n\r\n';
const TOUCH_THRESHOLD = 60; // should always be min(client_age) - max(client_age/2)

const isNormalInteger = (str) => {
  var n = Math.floor(Number(str));
  return String(n) === str && n >= 0;
};

// Logger interface for user to send log messages to server
const Logger = function(id, client) {
  this.id = id;
  this.client = client;

  this.info = (format, ...args) => this._write(CommonUtils.LoggingLevels.Info, 0, format, ...args);
  this.error = (format, ...args) => this._write(CommonUtils.LoggingLevels.Error, 0, format, ...args);
  this.debug = (format, ...args) => this._write(CommonUtils.LoggingLevels.Debug, 0, format, ...args);
  this.warn = (format, ...args) => this._write(CommonUtils.LoggingLevels.Warning, 0, format, ...args);
  this.trace = (format, ...args) => this._write(CommonUtils.LoggingLevels.Trace, 0, format, ...args);
  this.fatal = (format, ...args) => this._write(CommonUtils.LoggingLevels.Fatal, 0, format, ...args);
  this.verbose = (vlevel, format, ...args) => this._write(CommonUtils.LoggingLevels.Verbose, vlevel, format, ...args);

  //private members

  this._write = (level, vlevel, format, ...args) => client._sendLogRequest(level,
    this.id,
    this._logSources.getSourceFile(),
    this._logSources.getSourceLine(),
    this._logSources.getSourceFunc(),
    vlevel,
    undefined,
    format,
    ...args);

  this._logSources = {
    baseIndex: 6,
    getSourceFile: () => CommonUtils.getSourceFile(this._logSources.baseIndex),
    getSourceLine: () => CommonUtils.getSourceLine(this._logSources.baseIndex),
    getSourceFunc: () => CommonUtils.getSourceFunc(this._logSources.baseIndex),
  };
};

// Utility static functions
const Utils = {
  log: (m, ...args) => {
    console.log(m, ...args);
  },

  debugLog: (m, ...args) => {
    if (DEBUGGING) {
      console.log(`DEBUG: `, m, ...args);
    }
  },

  traceLog: (m, ...args) => {
    if (DEBUGGING) {
      console.log(`TRACE: `, m, ...args);
    }
  },

  vLog: (level, m, ...args) => {
    if (DEBUGGING && level <= VERBOSE_LEVEL) {
      console.log(`VERBOSE ${level}: `, m, ...args);
    }
  },

  hasFlag: (f, connection) => {
    if (connection === null) {
      return false;
    }
    return (connection.flags & f) !== 0;
  },

  // Encode Base64
  base64Encode: str => new Buffer(str).toString('base64'),

  base64Decode: encoded => new Buffer(encoded, 'base64').toString('utf-8'),

  getCipherAlgorithm: keyHex => `aes-${(keyHex.length / 2) * 8}-cbc`,

  encrypt: (request, connection) => {
    let encryptedRequest;
    try {
      let iv = new Buffer(crypto.randomBytes(16), 'hex');
      let cipher = crypto.createCipheriv(Utils.getCipherAlgorithm(connection.key), new Buffer(connection.key, 'hex'), iv);
      return iv.toString('hex') + ':' + connection.client_id + ':' + cipher.update(request, 'utf-8', 'base64') + cipher.final('base64') + PACKET_DELIMITER;
    } catch (err) {
      Utils.debugLog(err);
    }
    return '';
  },

  // Decrypt response from the server using symmetric key
  decrypt: (data, connection) => {
    if (connection === null) {
      return null;
    }
    try {
      const resp = data.split(':');
      const iv = resp[0];
      const clientId = resp.length === 3 ? resp[1] : '';
      const actualData = resp.length === 3 ? resp[2] : resp[1];
      const binaryData = new Buffer(actualData, 'base64');
      Utils.vLog(8, 'Reading ' + data.trim() + ' >>> parts: ' + iv + ' >>> ' + actualData.trim() + ' >>> ' + connection.key);
      let decipher = crypto.createDecipheriv(Utils.getCipherAlgorithm(connection.key), new Buffer(connection.key, 'hex'), new Buffer(iv, 'hex'));
      decipher.setAutoPadding(false);

      let plain = decipher.update(binaryData, 'base64', 'utf-8');
      plain += decipher.final('utf-8');
      // Remove non-ascii characters from decrypted text ! Argggh!
      plain = plain.replace(/[^A-Za-z 0-9 \.,\?""!@#\$%\^&\*\(\)-_=\+;:<>\/\\\|\}\{\[\]`~]*/g, '');
      return plain;
    } catch (err) {
      Utils.vLog(9, 'decrypt-error: ', err);
    }

    return null;
  },

  extractPublicKey: privateKey => new NodeRSA(privateKey.key).exportKey('public'),

  generateKeypair: (keySize) => {
    const key = new NodeRSA({
      b: keySize
    });
    key.setOptions({
      encryptionScheme: 'pkcs1'
    });
    Utils.debugLog('Key generated');
    return {
      privatePEM: key.exportKey('private'),
      publicPEM: key.exportKey('public'),
    };
  },

  // Decrypt response from the server using asymetric key
  decryptRSA: (response, privateKey) => {
    try {
      return crypto.privateDecrypt(privateKey, new Buffer(response.toString(), 'base64')).toString('utf-8');
    } catch (err) {
      Utils.log(err);
    }
    return null;
  },

  // Encrypts string using key
  encryptRSA: (str, publicKey) => {
    try {
      return crypto.publicEncrypt(publicKey, new Buffer(str, 'utf-8')).toString('base64');
    } catch (err) {
      Utils.log(err);
    }
    return null;
  },

  // Get current date in microseconds
  now: () => parseInt((new Date()).getTime() / 1000, 10),

  getTimestamp: () => Utils.now(),

  // Returns UTC time
  getCurrentTimeUTC: () => {
    const newDate = new Date();
    return newDate.getTime() + newDate.getTimezoneOffset() * 60000;
  },
};

const ResidueClient = function() {

  this._params = {
    // user provided options for seamless connection
    //   app, host, connect_port
    options: {},

    // connecting object containing:
    //   client_id, age, date_created, key, logging_port
    connection: null,

    // rsaKey is keypair object
    rsaKey: null,

    // serverRsaKey is keypair object
    serverRsaKey: null,

    // whether connected to the server or not
    connected: false,

    // Underlying sockets
    connectionSocket: new net.Socket(),
    loggingSocket: new net.Socket(),

    // Status for sockets
    isLoggingSocketConnected: false,

    // callbacks on specific occasions
    sendRequestBacklogCallbacks: [],
    loggingSocketCallbacks: [],

    // locks for mutex
    locks: {},
  };

  this._params.locks[this._params.connectionSocket.address().port] = false;
  this._params.locks[this._params.loggingSocket.address().port] = false;

  // Send request to the server
  // This function decides whether to back-log the request or dispatch it to
  // the server
  this._sendRequest = (request, socket, nolock /* = false */ , compress /* = false */ ) => {
    if (typeof nolock === 'undefined') {
      nolock = false;
    }
    if (typeof compress === 'undefined') {
      compress = false;
    }
    if (!nolock && this._params.locks[socket.address().port]) {
      this._params.sendRequestBacklogCallbacks.push(() => {
        Utils.debugLog('Sending request via callback');
        this._sendRequest(request, socket, false, compress);
      });
      return;
    }
    let finalRequest = JSON.stringify(request);
    if (compress) {
      finalRequest = new Buffer(zlib.deflateSync(finalRequest)).toString('base64');
    }
    const encryptedRequest = Utils.encrypt(finalRequest, this._params.connection);
    Utils.vLog(9, 'Payload (Plain): ', encryptedRequest);
    Utils.vLog(8, 'Locking ' + socket.address().port);
    this._params.locks[socket.address().port] = true;
    try {
      Utils.debugLog('Sending...');
      const self = this;
      socket.write(encryptedRequest, 'utf-8', () => {
        self._params.locks[socket.address().port] = false;
        Utils.vLog(8, 'Unlocking ' + socket.address().port);
        setTimeout(() => {
          if (self._params.sendRequestBacklogCallbacks.length > 0) {
            const cb = self._params.sendRequestBacklogCallbacks.splice(0, 1)[0];
            cb();
          }
        }, 10);
      });
    } catch (e) {
      Utils.vLog(8, 'Unlocking ' + socket.address().port + ' [because of exception]', e);
      this._params.locks[socket.address().port] = false;
      Utils.debugLog('Error while writing to socket...');
      Utils.debugLog(e);
    }
  };

  // Handle response from the server on connection requests
  this._params.connectionSocket.on('data', (data) => {
    let decryptedData = Utils.decrypt(data.toString(), this._params.connection);
    if (decryptedData === null) {
      decryptedData = Utils.decryptRSA(data, this._params.rsaKey.privateKey);
    }
    if (decryptedData === null) {
      Utils.log('Unable to read response: ' + data);
      return;
    }
    const dataJson = JSON.parse(decryptedData.toString());
    Utils.vLog(8, 'Connection: ', dataJson);
    if (dataJson.status === 0 && typeof dataJson.key !== 'undefined' && dataJson.ack === 0) {
      Utils.debugLog('Connecting to Residue Server...(ack)');

      // connection re-estabilished
      this._params.disconnected_by_remote = false;

      this._params.connection = dataJson;
      // Need to acknowledge
      const request = {
        _t: Utils.getTimestamp(),
        type: ConnectType.ACK,
        client_id: this._params.connection.client_id
      };
      this._sendRequest(request, this._params.connectionSocket, true);
    } else if (dataJson.status === 0 && typeof dataJson.key !== 'undefined' && dataJson.ack === 1) {
      Utils.debugLog('Estabilishing full connection...');
      this._params.connection = dataJson;
      this._params.connected = true;
      Utils.vLog(8, `Connection socket: ${this._params.connectionSocket.address().port}`);
      if (!this._params.isLoggingSocketConnected) {
        this._params.loggingSocket.connect(this._params.connection.logging_port, this._params.options.host, () => {
          Utils.log(`Connected to Residue (v${this._params.connection.server_info.version})!`);
          this._params.isLoggingSocketConnected = true;
          Utils.vLog(8, `Logging socket: ${this._params.loggingSocket.address().port}`);
          this._params.connecting = false;
          const callbackCounts = this._params.loggingSocketCallbacks.length;
          for (let idx = 0; idx < callbackCounts; ++idx) {
            const cb = this._params.loggingSocketCallbacks.splice(0, 1)[0];
            cb();
          }
        });
      } else {
        this._params.connecting = false;
        const callbackCounts = this._params.loggingSocketCallbacks.length;
        for (let idx = 0; idx < callbackCounts; ++idx) {
          // trigger all the pending callbacks from backlog
          this._params.loggingSocketCallbacks.splice(0, 1)[0]();
        }
      }
    } else {
      Utils.log('Error while connecting to server: ');
      Utils.log(dataJson);
      this._params.connecting = false;
    }
  });

  // Handle when connection is destroyed
  this._params.connectionSocket.on('close', () => {
    Utils.log('Remote connection closed!');
    if (this._params.connected) {
      this._params.disconnected_by_remote = true;
    }
    this.disconnect();
  });

  this._params.connectionSocket.on('error', (error) => {
    Utils.log('Error occurred while connecting to residue server');
    Utils.log(error);
  });

  // Handle destruction of connection to logging server
  this._params.loggingSocket.on('close', () => {
    this._params.isLoggingSocketConnected = false;
  });

  // Notice we do not have any handler for loggingSocket response
  // this is because that is async connection
  this._params.loggingSocket.on('data', () => {});

  this._shouldTouch = () => {
    if (!this._params.connected || this._params.connecting) {
      // Can't touch
      return false;
    }
    if (this._params.connection.age === 0) {
      // Always alive!
      return false;
    }
    return this._params.connection.age - (Utils.now() - this._params.connection.date_created) < TOUCH_THRESHOLD;
  };

  this._touch = () => {
    if (this._params.connected) {
      if (this._params.connecting) {
        Utils.debugLog('Still touching...');
        return;
      }
      if (this._isClientValid()) {
        Utils.debugLog('Touching...');
        const request = {
          _t: Utils.getTimestamp(),
          type: ConnectType.TOUCH,
          client_id: this._params.connection.client_id
        };
        this._sendRequest(request, this._params.connectionSocket);
        this._params.connecting = true;
      } else {
        Utils.log('Could not touch, client already dead ' + (this._params.connection.date_created + this._params.connection.age) + ' < ' + Utils.now());
      }
    }
  };

  this._isClientValid = () => {
    if (!this._params.connected) {
      return false;
    }
    if (this._params.connection.age == 0) {
      return true;
    }
    return this._params.connection.date_created + this._params.connection.age >= Utils.now();
  };

  // Send log request to the server. No response is expected
  this._sendLogRequest = (level, loggerId, sourceFile, sourceLine, sourceFunc, verboseLevel, logDatetime, format, ...args) => {
    let datetime = logDatetime;
    if (typeof datetime === 'undefined') {
      datetime = this._params.options.utc_time ? Utils.getCurrentTimeUTC() : new Date().getTime();
      if (this._params.options.time_offset) {
        datetime += (1000 * this._params.options.time_offset); // offset is in seconds
      }
    }
    if (this._params.connecting) {
      Utils.debugLog('Still connecting...');
      this._params.loggingSocketCallbacks.push(() => {
        this._sendLogRequest(level, loggerId, sourceFile, sourceLine, sourceFunc, verboseLevel, datetime, format, ...args);
      });
      return;
    }

    if (!this._params.connected) {
      Utils.log('Not connected to the server yet');
      if (this._params.disconnected_by_remote) {
        Utils.debugLog('Queueing...');
        this._params.loggingSocketCallbacks.push(() => {
          this._sendLogRequest(level, loggerId, sourceFile, sourceLine, sourceFunc, verboseLevel, datetime, format, ...args);
        });
        const totalListener = this._params.connectionSocket.listenerCount('connect');
        if (totalListener >= 1) {
          Utils.log('Checking for connection...' + totalListener);
          this._params.connectionSocket.emit('connect');
        } else {
          Utils.log('Retrying to connect...');
          connect(this._params.options);
        }
      }
      return;
    }

    Utils.debugLog('Checking health...[' + loggerId + ']');

    if (!this._isClientValid()) {
      Utils.debugLog('Resetting connection...');
      this._params.loggingSocketCallbacks.push(() => {
        Utils.debugLog('Sending log from log callback... [' + loggerId + ']');
        sendLogRequest(level, loggerId, sourceFile, sourceLine, sourceFunc, verboseLevel, datetime, format, ...args);
      });
      Utils.debugLog('Destroying connection socket');
      this._params.connectionSocket.destroy();
      this._params.loggingSocket.destroy();
      disconnect();
      connect(this._params.options);
      return;
    }

    if (this._shouldTouch()) {
      Utils.debugLog('Touching first...');
      this._params.loggingSocketCallbacks.push(() => {
        Utils.debugLog('Sending log from touch callback... [' + loggerId + ']');
        this._sendLogRequest(level, loggerId, sourceFile, sourceLine, sourceFunc, verboseLevel, datetime, format, ...args);
      });
      this._touch();
      return;
    }

    Utils.debugLog('Sending log request [' + loggerId + ']...');

    const fullMessage = CommonUtils.translateArgs(true, ...args);
    const request = {
      _t: Utils.getTimestamp(),
      datetime: datetime,
      logger: loggerId,
      msg: util.format(format, ...fullMessage),
      file: sourceFile,
      line: sourceLine,
      func: sourceFunc,
      app: this._params.options.application_id,
      level: level,
    };
    if (typeof verboseLevel !== 'undefined') {
      request.vlevel = verboseLevel;
    }
    this._sendRequest(request, this._params.loggingSocket, false, Utils.hasFlag(Flag.COMPRESSION, this._params.connection));
  };

  // public exported functions

  this.version = () => require('./../package.json').version;

  this.type = () => 'js';

  this.isConnected = () => this._params.connected;

  // Securily connect to residue server using defined options
  this.connect = (options) => {
    if (this._params.connected && this._params.connection !== null) {
      Utils.log('Already connected to the server with ID [' + this._params.connection.client_id + ']')
      return;
    }
    this._params.connecting = true;
    try {
      this._params.options = typeof options === 'undefined' ? this._params.options : options;
      // Normalize
      if (typeof this._params.options.url !== 'undefined') {
        const parts = this._params.options.url.split(':');
        if (parts.length < 2 || !isNormalInteger(parts[1])) {
          throw 'Invalid URL format for residue';
        }
        this._params.options.host = parts[0];
        this._params.options.connect_port = parseInt(parts[1]);
      }
      if (typeof this._params.options.client_id === 'undefined' &&
        typeof this._params.options.client_private_key === 'undefined') {
        // Generate new key for key-exchange
        const keySize = this._params.options.rsaKey_size || 2048;
        Utils.log('Generating ' + keySize + '-bit key...');
        const generatedKey = Utils.generateKeypair(keySize);
        this._params.rsaKey = {
          isGenerated: true,
          privateKey: {
            key: generatedKey.privatePEM,
            padding: crypto.constants.RSA_PKCS1_PADDING,
          },
          publicKey: {
            key: generatedKey.publicPEM,
            padding: crypto.constants.RSA_PKCS1_PADDING,
          }
        };
        Utils.debugLog('Key generated');
      } else {
        this._params.rsaKey = {
          generated: false,
          privateKey: {
            key: fs.readFileSync(path.resolve(this._params.options.client_private_key)).toString(),
            passphrase: this._params.options.client_key_secret ?
              new Buffer(this._params.options.client_key_secret, 'hex').toString('utf-8') : null,
            padding: crypto.constants.RSA_PKCS1_PADDING,
          },
          publicKey: {
            padding: crypto.constants.RSA_PKCS1_PADDING,
          }
        };
        if (typeof this._params.options.client_public_key !== 'undefined') {
          this._params.rsaKey.publicKey.key = fs.readFileSync(path.resolve(this._params.options.client_public_key)).toString();
        } else {
          if (this._params.rsaKey.privateKey.passphrase === null) {
            this._params.rsaKey.publicKey.key = Utils.extractPublicKey(this._params.rsaKey.privateKey);
          } else {
            throw 'ERROR: You specified client_key_secret and did not provide client_public_key. We cannot extract public-key for encrypted private keys. Please provide public key manually';
          }
        }
        Utils.vLog(8, 'Known client...');
      }
      if (typeof this._params.options.server_public_key !== 'undefined') {
        this._params.serverRsaKey = {
          publicKey: {
            key: fs.readFileSync(path.resolve(this._params.options.server_public_key)).toString(),
            padding: crypto.constants.RSA_PKCS1_PADDING,
          },
        };
      }
      Utils.log('Intializing connection...');
      this._params.connectionSocket.connect(this._params.options.connect_port, this._params.options.host, () => {
        let request = {
          _t: Utils.getTimestamp(),
          type: ConnectType.CONN,
        };
        if (this._params.rsaKey.isGenerated) {
          request.rsa_public_key = Utils.base64Encode(this._params.rsaKey.publicKey.key);
        } else {
          request.client_id = this._params.options.client_id;
        }
        let r = JSON.stringify(request);
        if (this._params.serverRsaKey !== null) {
          r = Utils.encryptRSA(r, this._params.serverRsaKey.publicKey);
        }
        const fullReq = r + PACKET_DELIMITER;
        this._params.connectionSocket.write(fullReq);
      });
    } catch (e) {
      Utils.log('Error occurred while connecting to residue server');
      Utils.log(e);
      this._params.connecting = false;
    }
  };

  // Disconnect from the server safely.
  this.disconnect = () => {
    Utils.traceLog('disconnect()');
    this._params.connected = false;
    this._params.connecting = false;
    this._params.connection = null;
    this._params.isLoggingSocketConnected = false;
    if (this._params.connected) {
      try {
        if (this._params.connectionSocket.destroyed) {
          Utils.log('Disconnecting gracefully...');
          this._params.loggingSocket.end();
        } else {
          Utils.log('Disconnecting...');
          // Following will call 'close' -> disconnect -> gracefully close
          this._params.connectionSocket.end();
        }
      } catch (err) {

      }
    }
  };

  this.loadConfiguration = (jsonOrFilename) => {
    const conf = CommonUtils.confJson(jsonOrFilename);
    if (conf === false) {
      console.error('Please select JSON or JSON filename that contains configurations');
      return false;
    }
    this._params.options = JSON.parse(conf);
    Utils.log('Configuration loaded');
    return true;
  };

  this.getLogger = id => new Logger(id, this);
};

module.exports = ResidueClient;
